### PR TITLE
Adds test file for program unit tests

### PR DESCRIPTION
**Pull request by josh146**
**Description of the Change:**

* Reorders the test suite, creates a new `frontend/test_program.py` file containing program unit tests
* Removes program specific unit tests from `test_engine.py`
* Fixes the `test_circuitdrawer.py` tests to work with the new program object

### CODE REVIEW

File: tests/frontend/test_program.py [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+class TestRegRefs:
+    """Testing register references."""
+
+    def test_corrupt_index(self, prog):
+        """User messes up the RegRef indices."""
+        with prog.context as q:
+            q[0].ind = 1
+            with pytest.raises(
+                program.RegRefError, match="RegRef state has become inconsistent"
+            ):
```
Comment (smite) [inline tests/frontend/test_program.py]: Could we increase the allowed line length to 100--120 chars to prevent black splitting up lines like this?

**Comment by smite** [comment]
Looks good, I'll add some more Engine-specific tests once I have the LocalEngine class done.

File: tests/frontend/test_program.py (hunk 1 above)
Comment (smite) [reply tests/frontend/test_program.py]: I mean this is not the worst formatting but I tend to use slightly longer lines myself...

Comment (josh146) [reply tests/frontend/test_program.py]: We should pick a standard Black line length to ensure that we are all using the same formatting, @co9olguy @trbromley @jswinarton

File: tests/frontend/test_circuitdrawer.py [hunk 2]
```diff
@@ -942,15 +941,15 @@ def test_compile(self, tmpdir):
...
 
-        with eng:
+        with prog.context as q:
             ops.Dgate(1) | (q[1])
             ops.Rgate(1) | (q[1])
             ops.S2gate(1) | (q[0], q[1])
 
         subdir = tmpdir.join("subdir")
-        document = eng.draw_circuit(print_queued_ops=True, tex_dir=subdir)[0]
+        document = prog.draw_circuit(tex_dir=subdir)[0]
```
Comment (co9olguy) [inline tests/frontend/test_circuitdrawer.py]: didn't realize before how comprehensive this file was. 1000 lines :)

File: tests/frontend/test_program.py [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+            D | r
+        assert cont.can_follow(prog)
+        assert prog.reg_refs == cont.init_reg_refs
+        assert prog.unused_indices == cont.init_unused_indices
+
+    def test_print_commands(self, eng, prog):
+        """Program.print and Engine.print_applied return correct strings."""
+        prog = sf.Program(2)
+        res = []
+        print_fn = lambda x: res.append(x.__str__())
```
Comment (co9olguy) [inline tests/frontend/test_program.py]: our usage of `res` and `print_fn` is a bit confusing to follow in the tests; it's not easy to follow the flow. Perhaps some brief explanatory comment could be useful here

File: tests/frontend/test_program.py (hunk 1 above)
Comment (co9olguy) [reply tests/frontend/test_program.py]: go ahead and just pick one

Comment (josh146) [reply tests/frontend/test_program.py]: Can I make an executive decision and say 100 chars per line

File: tests/frontend/test_program.py (hunk 3 above)
Comment (josh146) [reply tests/frontend/test_program.py]: done

**Pull request merged**